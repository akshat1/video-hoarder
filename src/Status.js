/**
 * @enum {string} Status
 */
export default {
  Pending: 'pending',
  Running: 'running',
  Failed: 'failed',
  Succeeded: 'succeeded'
};
