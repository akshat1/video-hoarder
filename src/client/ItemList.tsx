import { Mutation, Query, Subscription } from "./gql";
import { JobsQueryResponse } from "./gql/job";
import { Item } from "./Item";
import { useMutation, useQuery, useSubscription } from "@apollo/client";
import { Delete } from "@mui/icons-material";
import { Button, CircularProgress, Grid } from "@mui/material";
import _ from "lodash";
import React, { FunctionComponent } from "react";

export const ItemList:FunctionComponent = () => {
  const {
    data,
    loading: fetchingJobs,
    error: jobFetchingError,
    refetch: refetchJobs,
  } = useQuery<JobsQueryResponse>(Query.Jobs);
  const jobs = _.get(data, "jobs", []);

  const [doCleanUp, cleanUpThunk] = useMutation(Mutation.RemoveAllDoneJobs, {
    refetchQueries: [{ query: Query.Jobs }, "Jobs"],
  });

  const handleCleanUpClicked = () => doCleanUp();

  useSubscription(Subscription.JobAdded, {
    onSubscriptionData: () => refetchJobs(),
  });

  useSubscription(Subscription.JobRemoved, {
    onSubscriptionData: () => refetchJobs(),
  });

  useSubscription(Subscription.JobUpdated);

  let loadingElement = null;
  let errorElement = null;
  let jobElements = [];

  if (fetchingJobs) 
    loadingElement = <>Loading</>;

  if (jobFetchingError) 
    errorElement = <>Error</>;

  if (!(fetchingJobs && jobFetchingError)) 
    jobElements = jobs.map(j =>
      <Grid item xs={12} key={j.id}>
        <Item job={j}/>
      </Grid>
    );
  
  const cleanUpIcon = cleanUpThunk?.loading ? <CircularProgress /> : <Delete />;

  return (
    <Grid container spacing={2}>
      <Grid item xs={12}>
        <Button startIcon={cleanUpIcon} variant="outlined" onClick={handleCleanUpClicked} disabled={cleanUpThunk?.loading}>
          Clean-up
        </Button>
      </Grid>
      {loadingElement}
      {errorElement}
      {jobElements}
    </Grid>
  );
};
