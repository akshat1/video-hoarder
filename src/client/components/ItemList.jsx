import Item from "./Item.jsx";
import Style from "./ItemList.less";
import PropTypes from "prop-types";
import React from "react";

const ItemList = ({ items }) => (
  <div className={Style.wrapper}>
    <For each="item" of={items}>
      <Item item={item} key={item.id}/>
    </For>
  </div>
);

ItemList.propTypes = {
  items: PropTypes.arrayOf(PropTypes.shape({
    addedAt: PropTypes.string.isRequired,
    id: PropTypes.string.isRequired,
    status: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired,
    updatedAt: PropTypes.string.isRequired,
    url: PropTypes.string.isRequired,
  })),
};

ItemList.defaultProps = {
  items: [],
};

export default ItemList;
