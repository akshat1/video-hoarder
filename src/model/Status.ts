export enum Status {
  Failed = "failed",
  Paused = "paused",
  Pending = "pending",
  Running = "running",
  Succeeded = "succeeded",
}

export enum StatusFilterValue {
  All = "all",
  Failed = "failed",
  Paused = "paused",
  Pending = "pending",
  Running = "running",
  Succeeded = "succeeded",
}


export const hasStarted = (status: Status): boolean =>
  status === Status.Running || status === Status.Failed || status === Status.Succeeded || status === Status.Paused;

export const hasConcluded = (status: Status): boolean =>
  status === Status.Failed || status === Status.Succeeded;

export const isPending = (status: Status): boolean =>
  status === Status.Pending;
