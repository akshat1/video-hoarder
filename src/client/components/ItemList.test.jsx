import React from 'react';
import { shallow } from 'enzyme';
import ItemList from './ItemList.jsx';
import { Status } from '../../Status';
import { getFakeItem } from '../fixtures/item';

describe('components/ItemList', () => {
  test('ItemList should match snapshot', () => {
    const items = [
      getFakeItem(Status.Succeeded),
      getFakeItem(Status.Succeeded),
      getFakeItem(Status.Failed),
      getFakeItem(Status.Failed),
      getFakeItem(Status.Running),
      getFakeItem(Status.Running),
      getFakeItem(Status.Running),
      getFakeItem(Status.Running),
      getFakeItem(Status.Succeeded),
      getFakeItem(Status.Running),
      getFakeItem(Status.Succeeded),
      getFakeItem(Status.Running),
      getFakeItem(Status.Pending),
      getFakeItem(Status.Running),
      getFakeItem(Status.Pending),
      getFakeItem(Status.Pending),
      getFakeItem(Status.Pending),
    ];

    expect(shallow(<ItemList items={items} />)).toMatchSnapshot()
  });
});
