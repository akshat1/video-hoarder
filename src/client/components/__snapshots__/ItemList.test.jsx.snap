// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/ItemList ItemList should match snapshot 1`] = `
<div
  className="wrapper"
>
  <Item
    item={
      Object {
        "addedAt": "2020-03-29T14:30:45.000Z",
        "id": "3idh834dh348dhhfufh48",
        "status": "succeeded",
        "title": "A collectible video",
        "updatedAt": "Sun Mar 29 2020 14:32:30 GMT+0000 (Coordinated Universal Time)",
        "url": "https://www.youtube.com/watch?v=eQ5Ru7Zu_1I?foo=bar&baz=qux",
      }
    }
    key="3idh834dh348dhhfufh48"
  />
  <Item
    item={
      Object {
        "addedAt": "2020-03-29T14:30:45.000Z",
        "id": "3idh834dh348dhhfufh48",
        "status": "succeeded",
        "title": "A collectible video",
        "updatedAt": "Sun Mar 29 2020 14:32:30 GMT+0000 (Coordinated Universal Time)",
        "url": "https://www.youtube.com/watch?v=eQ5Ru7Zu_1I?foo=bar&baz=qux",
      }
    }
    key="3idh834dh348dhhfufh48"
  />
  <Item
    item={
      Object {
        "addedAt": "2020-03-29T14:30:45.000Z",
        "id": "3idh834dh348dhhfufh48",
        "status": "failed",
        "title": "A collectible video",
        "updatedAt": "Sun Mar 29 2020 14:32:30 GMT+0000 (Coordinated Universal Time)",
        "url": "https://www.youtube.com/watch?v=eQ5Ru7Zu_1I?foo=bar&baz=qux",
      }
    }
    key="3idh834dh348dhhfufh48"
  />
  <Item
    item={
      Object {
        "addedAt": "2020-03-29T14:30:45.000Z",
        "id": "3idh834dh348dhhfufh48",
        "status": "failed",
        "title": "A collectible video",
        "updatedAt": "Sun Mar 29 2020 14:32:30 GMT+0000 (Coordinated Universal Time)",
        "url": "https://www.youtube.com/watch?v=eQ5Ru7Zu_1I?foo=bar&baz=qux",
      }
    }
    key="3idh834dh348dhhfufh48"
  />
  <Item
    item={
      Object {
        "addedAt": "2020-03-29T14:30:45.000Z",
        "id": "3idh834dh348dhhfufh48",
        "status": "running",
        "title": "A collectible video",
        "updatedAt": "Sun Mar 29 2020 14:32:30 GMT+0000 (Coordinated Universal Time)",
        "url": "https://www.youtube.com/watch?v=eQ5Ru7Zu_1I?foo=bar&baz=qux",
      }
    }
    key="3idh834dh348dhhfufh48"
  />
  <Item
    item={
      Object {
        "addedAt": "2020-03-29T14:30:45.000Z",
        "id": "3idh834dh348dhhfufh48",
        "status": "running",
        "title": "A collectible video",
        "updatedAt": "Sun Mar 29 2020 14:32:30 GMT+0000 (Coordinated Universal Time)",
        "url": "https://www.youtube.com/watch?v=eQ5Ru7Zu_1I?foo=bar&baz=qux",
      }
    }
    key="3idh834dh348dhhfufh48"
  />
  <Item
    item={
      Object {
        "addedAt": "2020-03-29T14:30:45.000Z",
        "id": "3idh834dh348dhhfufh48",
        "status": "running",
        "title": "A collectible video",
        "updatedAt": "Sun Mar 29 2020 14:32:30 GMT+0000 (Coordinated Universal Time)",
        "url": "https://www.youtube.com/watch?v=eQ5Ru7Zu_1I?foo=bar&baz=qux",
      }
    }
    key="3idh834dh348dhhfufh48"
  />
  <Item
    item={
      Object {
        "addedAt": "2020-03-29T14:30:45.000Z",
        "id": "3idh834dh348dhhfufh48",
        "status": "running",
        "title": "A collectible video",
        "updatedAt": "Sun Mar 29 2020 14:32:30 GMT+0000 (Coordinated Universal Time)",
        "url": "https://www.youtube.com/watch?v=eQ5Ru7Zu_1I?foo=bar&baz=qux",
      }
    }
    key="3idh834dh348dhhfufh48"
  />
  <Item
    item={
      Object {
        "addedAt": "2020-03-29T14:30:45.000Z",
        "id": "3idh834dh348dhhfufh48",
        "status": "succeeded",
        "title": "A collectible video",
        "updatedAt": "Sun Mar 29 2020 14:32:30 GMT+0000 (Coordinated Universal Time)",
        "url": "https://www.youtube.com/watch?v=eQ5Ru7Zu_1I?foo=bar&baz=qux",
      }
    }
    key="3idh834dh348dhhfufh48"
  />
  <Item
    item={
      Object {
        "addedAt": "2020-03-29T14:30:45.000Z",
        "id": "3idh834dh348dhhfufh48",
        "status": "running",
        "title": "A collectible video",
        "updatedAt": "Sun Mar 29 2020 14:32:30 GMT+0000 (Coordinated Universal Time)",
        "url": "https://www.youtube.com/watch?v=eQ5Ru7Zu_1I?foo=bar&baz=qux",
      }
    }
    key="3idh834dh348dhhfufh48"
  />
  <Item
    item={
      Object {
        "addedAt": "2020-03-29T14:30:45.000Z",
        "id": "3idh834dh348dhhfufh48",
        "status": "succeeded",
        "title": "A collectible video",
        "updatedAt": "Sun Mar 29 2020 14:32:30 GMT+0000 (Coordinated Universal Time)",
        "url": "https://www.youtube.com/watch?v=eQ5Ru7Zu_1I?foo=bar&baz=qux",
      }
    }
    key="3idh834dh348dhhfufh48"
  />
  <Item
    item={
      Object {
        "addedAt": "2020-03-29T14:30:45.000Z",
        "id": "3idh834dh348dhhfufh48",
        "status": "running",
        "title": "A collectible video",
        "updatedAt": "Sun Mar 29 2020 14:32:30 GMT+0000 (Coordinated Universal Time)",
        "url": "https://www.youtube.com/watch?v=eQ5Ru7Zu_1I?foo=bar&baz=qux",
      }
    }
    key="3idh834dh348dhhfufh48"
  />
  <Item
    item={
      Object {
        "addedAt": "2020-03-29T14:30:45.000Z",
        "id": "3idh834dh348dhhfufh48",
        "status": "pending",
        "title": "A collectible video",
        "updatedAt": "Sun Mar 29 2020 14:32:30 GMT+0000 (Coordinated Universal Time)",
        "url": "https://www.youtube.com/watch?v=eQ5Ru7Zu_1I?foo=bar&baz=qux",
      }
    }
    key="3idh834dh348dhhfufh48"
  />
  <Item
    item={
      Object {
        "addedAt": "2020-03-29T14:30:45.000Z",
        "id": "3idh834dh348dhhfufh48",
        "status": "running",
        "title": "A collectible video",
        "updatedAt": "Sun Mar 29 2020 14:32:30 GMT+0000 (Coordinated Universal Time)",
        "url": "https://www.youtube.com/watch?v=eQ5Ru7Zu_1I?foo=bar&baz=qux",
      }
    }
    key="3idh834dh348dhhfufh48"
  />
  <Item
    item={
      Object {
        "addedAt": "2020-03-29T14:30:45.000Z",
        "id": "3idh834dh348dhhfufh48",
        "status": "pending",
        "title": "A collectible video",
        "updatedAt": "Sun Mar 29 2020 14:32:30 GMT+0000 (Coordinated Universal Time)",
        "url": "https://www.youtube.com/watch?v=eQ5Ru7Zu_1I?foo=bar&baz=qux",
      }
    }
    key="3idh834dh348dhhfufh48"
  />
  <Item
    item={
      Object {
        "addedAt": "2020-03-29T14:30:45.000Z",
        "id": "3idh834dh348dhhfufh48",
        "status": "pending",
        "title": "A collectible video",
        "updatedAt": "Sun Mar 29 2020 14:32:30 GMT+0000 (Coordinated Universal Time)",
        "url": "https://www.youtube.com/watch?v=eQ5Ru7Zu_1I?foo=bar&baz=qux",
      }
    }
    key="3idh834dh348dhhfufh48"
  />
  <Item
    item={
      Object {
        "addedAt": "2020-03-29T14:30:45.000Z",
        "id": "3idh834dh348dhhfufh48",
        "status": "pending",
        "title": "A collectible video",
        "updatedAt": "Sun Mar 29 2020 14:32:30 GMT+0000 (Coordinated Universal Time)",
        "url": "https://www.youtube.com/watch?v=eQ5Ru7Zu_1I?foo=bar&baz=qux",
      }
    }
    key="3idh834dh348dhhfufh48"
  />
</div>
`;
